File: 94-binary-inorder/go/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderTraversal(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	left := []*TreeNode{root}
	right := []*TreeNode{}
	res := []int{}

	for len(left) != 0 || len(right) != 0 {
		if len(left) == 0 {
			top := right[len(right)-1]
			right = right[:len(right)-1]
			res = append(res, top.Val)
			if top.Right != nil {
				left = append(left, top.Right)
			}
		} else {
			top := left[len(left)-1]
			left = left[:len(left)-1]
			right = append(right, top)
			if top.Left != nil {
				left = append(left, top.Left)
			}
		}
	}

	return res
}

File: 94-binary-inorder/cpp/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
 public:
  vector<int> inorderTraversal(TreeNode* root) {
    if (root == NULL) {
      return {};
    }
    stack<TreeNode*> left, right;
    left.push(root);
    vector<int> res;

    while (!left.empty() || !right.empty()) {
      if (left.empty()) {
        res.push_back(right.top()->val);
        if (right.top()->right != NULL) {
          left.push(right.top()->right);
        }
        right.pop();
      } else {
        right.push(left.top());
        left.pop();
        if (right.top()->left != NULL) {
          left.push(right.top()->left);
        }
      }
    }

    return res;
  }
};